src/storage/mvcc/reader/backward_scanner.rs

```diff
@@ -0,0 +1,713 @@
+pub struct BackwardScanner<S: Snapshot> {
+    snapshot: S,
+    fill_cache: bool,
+    omit_value: bool,
+    isolation_level: IsolationLevel,
+
+    /// `lower_bound` and `upper_bound` is only used to create `default_cursor`. It will be
+    /// consumed after `default_cursor` is being created.
+    lower_bound: Option<Vec<u8>>,
+    upper_bound: Option<Vec<u8>>,
+
+    ts: u64,
+
+    lock_cursor: Cursor<S::Iter>,
+    write_cursor: Cursor<S::Iter>,
+
+    /// `default cursor` is lazy created only when it's needed.
+    default_cursor: Option<Cursor<S::Iter>>,
+
+    /// Is iteration started
+    is_started: bool,
+
+    statistics: Statistics,
+}
+
+impl<S: Snapshot> BackwardScanner<S> {
+    /// Take out and reset the statistics collected so far.
+    pub fn take_statistics(&mut self) -> Statistics {
+        ::std::mem::replace(&mut self.statistics, Statistics::default())
+    }
+
+    /// Get the next key-value pair, in backward order.
+    pub fn read_next(&mut self) -> Result<Option<(Key, Value)>> {
+        if !self.is_started {
+            self.write_cursor.seek_to_last(&mut self.statistics.write);
+            self.lock_cursor.seek_to_last(&mut self.statistics.lock);
+            self.is_started = true;
+        }
+
+        // Similar to forward scanner, the general idea is to simultaneously step write
+        // cursor and lock cursor. Please refer to `ForwardScanner` for details.
+
+        loop {
+            let (current_user_key, has_write, has_lock) = {
+                let w_key = if self.write_cursor.valid() {
+                    Some(self.write_cursor.key(&mut self.statistics.write))
+                } else {
+                    None
+                };
+                let l_key = if self.lock_cursor.valid() {
+                    Some(self.lock_cursor.key(&mut self.statistics.lock))
+                } else {
+                    None
+                };
+                match (w_key, l_key) {
+                    (None, None) => return Ok(None),
+                    (None, Some(lk)) => (lk.to_vec(), false, true),
+                    (Some(wk), None) => (Key::truncate_ts_for(wk)?.to_vec(), true, false),
+                    (Some(wk), Some(lk)) => match Key::truncate_ts_for(wk)?.cmp(lk) {
+                        Ordering::Less => {
+                            // We are scanning from largest user key to smallest user key, so this
+                            // indicate that we meet a lock first, thus its corresponding write
+                            // does not exist.
+                            (lk.to_vec(), false, true)
+                        }
+                        Ordering::Greater => {
+                            // We meet write first, so the lock of the write key does not exist.
+                            (Key::truncate_ts_for(wk)?.to_vec(), true, false)
+                        }
+                        Ordering::Equal => (Key::truncate_ts_for(wk)?.to_vec(), true, true),
```

Comment:
We don't need to call truncate_ts_for twice.

Reply:
done